mongo/session.go

```diff
@@ -193,6 +193,10 @@ func (s *sessionImpl) WithTransaction(ctx context.Context, fn func(sessCtx Sessi
 			default:
 			}
 
+			// End if context has timed out or been canceled, as retrying has no chance of success.
+			if IsDeadlineExceededError(err) {
```

Comment:
The check in the commit loop below checks: `ctx.Err() != nil`. IIUC we want to check for the same conditions. I think these should be consistent.

The documentation for [context.Err](https://golang.org/pkg/context/#Context) suggests that `ctx.Err()` returns a `DeadlineExceeded` or `Canceled` error.

I think the only slight difference between the two approaches is that a user's custom `Context` implementation could have `Err` return a different error. The broader `ctx.Err() != nil` check would always abort the retry loop. My guess is that is a rare use case. But if we make no assumptions about the custom error, maybe we should rely on the error labels if `ctx.Err` is not `DeadlineExceeded` or `Canceled`.

Reply:
The setup of `context`s in this loop is different than the `CommitLoop`. There are three different contexts present here. `ctx` is the context of `WithTransaction`. On L182, the context used in the execution of the function is a `SessionContext` based off of `ctx`. In the new tests, this second context is called `sessCtx`. Yet another context `c` is made based off of `sessCtx` with a short timeout (in the case of the "expired" test). `c` will have an error (`c.Err() != nil`).  The other contexts `sessCtx` and `ctx` will not.

I think `c` is not reachable from the code here. `ctx.Err()` and `sessCtx.Err()` can both equal nil even when there's been a timeout in `c`. I think we have to check  and unwrap the resulting `err` to look for `context.DeadlineExceeded` or `context.Canceled`.

Reply:
A great point, thanks for such a detailed response.

You're right; I don't think we can say a timeout on `ctx3` is not worth a retry (for the exact case you pointed out). The only thing we can really check is that `ctx.Err() != nil`, as `WithTransaction` has no chance of succeeding only if `ctx1` has expired or been canceled.

Removed the error helper, modified the tests and added a new one for a slow operation before `commitTransaction`. Let me know what you think.